src/emd/rt_bse.F

```diff
@@ -0,0 +1,1416 @@
+         CALL cp_cfm_scale_and_add(scale_factor, workspace, -scale_factor, rho_old(j))
+         ! Now, get the relevant number
+         partial_metric(j) = cp_cfm_norm(workspace, 'M')
+      END DO
+      metric = 0.0_dp
+      ! For more than one spin, do Cartesian sum of the different spin norms
+      DO j = 1, nspin
+         metric = metric + partial_metric(j)*partial_metric(j)
+      END DO
+      metric = SQRT(metric)
+      ! Deallocate workspace
+      IF (.NOT. PRESENT(workspace_opt)) CALL cp_cfm_release(workspace)
+      DEALLOCATE (partial_metric)
+   END FUNCTION
+
+! **************************************************************************************************
+!> \brief Determines the metric of the antihermitian part of the matrix
+!> \param real_fm Real part of the full matrix
+!> \param imag_fm Imaginary part of the full matrix
+! **************************************************************************************************
+   SUBROUTINE antiherm_metric(real_fm, imag_fm, workspace, metric)
+      TYPE(cp_fm_type), INTENT(IN)                      :: real_fm
+      TYPE(cp_fm_type), INTENT(IN), OPTIONAL            :: imag_fm
+      REAL(kind=dp), INTENT(OUT)                        :: metric
+      TYPE(cp_cfm_type), DIMENSION(:), POINTER          :: workspace
+      COMPLEX(kind=dp)                                  :: complex_one
+
+      ! Get the complex and complex conjugate matrix
+      IF (PRESENT(imag_fm)) THEN
+         CALL cp_fm_to_cfm(real_fm, imag_fm, workspace(1))
+      ELSE
+         CALL cp_fm_to_cfm(msourcer=real_fm, mtarget=workspace(1))
+      END IF
+      CALL cp_cfm_transpose(workspace(1), "C", workspace(2))
+      ! Subtract these, and get the metric
+      complex_one = CMPLX(1.0, 0.0, kind=dp)
+      CALL cp_cfm_scale_and_add(complex_one, workspace(1), -complex_one, workspace(2))
+      metric = cp_cfm_norm(workspace(1), "M")
+   END SUBROUTINE
+
+! **************************************************************************************************
+!> \brief For Taylor and Exact exp_method, calculates the matrix exponential of the
+!>        effective Hamiltonian. For BCH, calculates just the effective Hamiltonian. For other methods,
+!>        aborts the execution, as they are not implemented yet.
+!> \param rtbse_env Entry point of the calculation. Uses rho_workspace for Taylor and BCH. For exact,
+!>                  uses complex_workspace, complex_ham, complex_s, real_eigvals and exp_eigvals.
+!>                  Results are stored in ham_workspace.
+! **************************************************************************************************
+   SUBROUTINE ham_to_exp(rtbse_env)
+      TYPE(rtbse_env_type)                               :: rtbse_env
+      CHARACTER(len=*), PARAMETER                        :: routineN = "ham_to_exp"
+      INTEGER                                            :: j, handle
+      CALL timeset(routineN, handle)
+      DO, j = 1, rtbse_env%n_spin
```

Comment:
Remove the comma after the DO

Reply:
Thank you for spotting this, I removed the comma.